libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt

```diff
@@ -2,68 +2,66 @@
- * always find the first upcoming rest stop from the list of upcoming rest stops if available.
- * - returns null if route has no rest stop: In this case the API will return null
- * - returns [RestStopFromRoadObject] if route has 1 rest stop and hasn't been passed/reached
- * - returns null if route has 1 rest stop and it has already been passed/reached
- * - returns the first [RestStopFromRoadObject] that hasn't been passed/reached if route has multiple
- *   rest stops and you have passed a few of them
- * - returns null if route has multiple rest stops and all have been passed
- * This is recommended to be used with
- * - [RouteProgress] during Active Navigation.
- */
-fun List<UpcomingRoadObject>.getFirstUpcomingRestStop(): RestStopFromRoadObject? {
-    val restStops = this.getAllRestStops()
-    if (restStops.isNotEmpty()) {
-        val distanceToStart = restStops.first().distanceToStart
+fun RouteProgress.getFirstUpcomingRoadObject(@RoadObjectType.Type type: Int): UpcomingRoadObject? {
+    return ifNonNull(this.upcomingRoadObjects.firstOrNull { it.roadObject.objectType == type }) {
+        val distanceToStart = it.distanceToStart
         if (distanceToStart != null && distanceToStart > 0) {
-            return restStops.first()
+            it
+        } else {
+            null
         }
     }
-    return null
 }
 
 /**
- * Use the API to retrieve the list of all rest stops using [NavigationRoute].
- * This is recommended to be used with:
- * - [NavigationRoute] during Route Preview
+ * The function returns a list of all [UpcomingRoadObject] of type [clazz] present in a
+ * [DirectionsRoute] mapped to a [RouteLeg] in case there are multiple legs in a [DirectionsRoute].
+ *
+ * @param type determines [RoadObjectType] to filter. Any other value will return list of
+ * [UpcomingRoadObject] as null.
  */
-fun NavigationRoute.getAllRestStops(): RoadObjectExtra<RestStopFromIntersection> {
+fun NavigationRoute.getAllRoadObjects(@RoadObjectType.Type type: Int): RoadObjectExtra {
     val legIndexToRestStops = ifNonNull(this.directionsRoute.legs()) { routeLegs ->
-        val legToRestStops = hashMapOf<Int, List<RestStopFromIntersection>>()
-        routeLegs.forEachIndexed { legIndex, routeLeg ->
-            legToRestStops[legIndex] = ifNonNull(routeLeg.steps()) { steps ->
-                val restStops = mutableListOf<RestStopFromIntersection>()
-                steps.forEach { legStep ->
-                    restStops.addAll(
-                        RestStopFactory.createWithStepIntersection(
-                            legStep.intersections()
-                        )
-                    )
-                }
-                restStops
-            } ?: emptyList()
+        val legToRestStops = hashMapOf<Int, List<UpcomingRoadObject>>()
+        val restStops = mutableListOf<UpcomingRoadObject>()
+        routeLegs.forEachIndexed { legIndex, _ ->
+            restStops.addAll(this.nativeRoute().getAllRoadObjects(type))
```

Comment:
```
this.nativeRoute().getAllRoadObjects(type)
```
does not use a routeLeg, so it returns the same list for all legs, how's that supposed to work?

Reply:
I didn't get either. It would be nice to see some unit test for this filtering to understand a different cases the function covers.